### PR TITLE
Enable CA1852/CA1812 Analyzers

### LINKED ISSUE #7879: Enable `CA1852`/`CA1812` Analyzers
### Is your feature request related to a problem? Please describe

- `CA1852` Analyzer for sealing non-public classes which are not inherited from for performance reasons https://devblogs.microsoft.com/dotnet/performance_improvements_in_net_7/#analyzers
- `CA1812` Avoid uninstantiated internal classes for performance https://learn.microsoft.com/en-au/dotnet/fundamentals/code-analysis/quality-rules/ca1812

The configs are stored here: https://github.com/dotnet/winforms/blob/main/eng/CodeAnalysis.src.globalconfig.

### Describe the solution you'd like and alternatives you've considered

N/A

### Will this feature affect UI controls?

N/A

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #7879

Singular warning for CA1812 supressed due to usage in interop.

Synced globalconfig with runtime and added #TODO to items that were warning.
```
Line  182: dotnet_diagnostic.CA1416.severity = none # TODO: warning
Line  260: dotnet_diagnostic.CA1727.severity = suggestion # TODO: warning
Line  273: dotnet_diagnostic.CA1810.severity = none # TODO: warning
Line  297: dotnet_diagnostic.CA1821.severity = none # TODO: warning
Line  394: dotnet_diagnostic.CA1853.severity =  none # TODO: warning
Line  397: dotnet_diagnostic.CA1854.severity =  none # TODO: warning
Line  406: dotnet_diagnostic.CA2007.severity = none # TODO: warning
Line  439: dotnet_diagnostic.CA2019.severity = none # TODO: warning
Line  466: dotnet_diagnostic.CA2208.severity = none # TODO: warning
Line  533: dotnet_diagnostic.CA2245.severity = none # TODO: warning
Line  557: dotnet_diagnostic.CA2253.severity = none # TODO: warning
Line  851: dotnet_diagnostic.IL3000.severity = none # TODO: warning
Line  873: dotnet_diagnostic.SA1001.severity = none # TODO: warning
Line  912: dotnet_diagnostic.SA1014.severity = none # TODO: warning
Line  924: dotnet_diagnostic.SA1018.severity = none # TODO: warning
Line  930: dotnet_diagnostic.SA1020.severity = none # TODO: warning
Line  948: dotnet_diagnostic.SA1026.severity = none # TODO: warning
Line  951: dotnet_diagnostic.SA1027.severity = none # TODO: warning
Line  954: dotnet_diagnostic.SA1028.severity = none # TODO: warning
Line  963: dotnet_diagnostic.SA1102.severity = none # TODO: warning
Line  966: dotnet_diagnostic.SA1105.severity = none # TODO: warning
Line  996: dotnet_diagnostic.SA1113.severity = none # TODO: warning
Line 1002: dotnet_diagnostic.SA1115.severity = none # TODO: warning
Line 1020: dotnet_diagnostic.SA1121.severity = none # TODO: warning
Line 1044: dotnet_diagnostic.SA1129.severity = none # TODO: warning
Line 1074: dotnet_diagnostic.SA1141.severity = none # TODO: warning
Line 1077: dotnet_diagnostic.SA1142.severity = none # TODO: warning
Line 1095: dotnet_diagnostic.SA1205.severity = none # TODO: warning
Line 1116: dotnet_diagnostic.SA1212.severity = none # TODO: warning
Line 1137: dotnet_diagnostic.SA1302.severity = none # TODO: warning
Line 1179: dotnet_diagnostic.SA1400.severity = none # TODO: warning
Line 1191: dotnet_diagnostic.SA1404.severity = none # TODO: warning
Line 1203: dotnet_diagnostic.SA1408.severity = none # TODO: warning
Line 1212: dotnet_diagnostic.SA1411.severity = none # TODO: warning
Line 1278: dotnet_diagnostic.SA1518.severity = none # TODO: warning
Line 1494: dotnet_diagnostic.IDE0020.severity = none # TODO: warning
Line 1521: dotnet_diagnostic.IDE0029.severity = none # TODO: warning
Line 1527: dotnet_diagnostic.IDE0031.severity = none # TODO: warning
Line 1542: dotnet_diagnostic.IDE0036.severity = none # TODO: warning
Line 1593: dotnet_diagnostic.IDE0054.severity = none # TODO: warning
Line 1608: dotnet_diagnostic.IDE0059.severity = none # TODO: warning
Line 1618: dotnet_diagnostic.IDE0062.severity = none # TODO: warning
Line 1627: dotnet_diagnostic.IDE0065.severity = none # TODO: warning
Line 1645: dotnet_diagnostic.IDE0074.severity = none # TODO: warning
Line 1669: dotnet_diagnostic.IDE0082.severity = none # TODO: warning
Line 1681: dotnet_diagnostic.IDE0100.severity = none # TODO: warning
Line 1684: dotnet_diagnostic.IDE0110.severity = none # TODO: warning
Line 1705: dotnet_diagnostic.IDE0170.severity = none # TODO: warning
Line 1711: dotnet_diagnostic.IDE0200.severity = none # TODO: warning
```

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7881)

### CODE REVIEW

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Were you able to confirm the analyzers flag the types you identified in https://github.com/dotnet/winforms/pull/7831#discussion_r982967549?

**Comment by elachlan** [comment]
> Were you able to confirm the analyzers flag the types you identified in [#7831 (comment)](https://github.com/dotnet/winforms/pull/7831#discussion_r982967549)?

It did not! I am looking into it.

**Comment by elachlan** [comment]
I think its the version of `Microsoft.CodeAnalysis.NetAnalyzers`, but I can't see it in `eng\Versions.props`. I guess this is tied to the SDK version?

**Comment by gpetrou** [comment]
> I think its the version of `Microsoft.CodeAnalysis.NetAnalyzers`, but I can't see it in `eng\Versions.props`. I guess this is tied to the SDK version?

The version that we use is in https://github.com/dotnet/winforms/blob/be7b0368d986b6220d259f2dd0785ccf3c728f0e/eng/CodeStyle.targets.
Usually when we update the version, we also update the rules config file to match what runtime has in https://github.com/dotnet/runtime/blob/release/7.0/eng/CodeAnalysis.src.globalconfig.

**Comment by elachlan** [comment]
I can sync with runtime if you want?

**Comment by gpetrou** [comment]
> I can sync with runtime if you want? Or did the team want to do that?

I am an external contributor, so I am not representing the team. What I did in the past was to sync the rules with runtime and add `severity = none` with TODOs for items that give warnings and can be done in separate PRs. You can see in the current main that some TODOs are still there.

**Comment by elachlan** [comment]
@NewellClark winforms is looking at implementing `CA1852`. Some of the classes are not being picked up by the analyzer. Am I misunderstanding the scope of the analyzer?

Example:
https://github.com/dotnet/winforms/blob/33342db8f9c2077e6ec86d7ca4e277821b20a609/src/System.Windows.Forms.Design/src/System/Drawing/Design/ContentAlignmentEditor.ContentUI.cs#L14-L17

**Comment by RussKie** [comment]
> The version that we use is in [`be7b036`/eng/CodeStyle.targets](https://github.com/dotnet/winforms/blob/be7b0368d986b6220d259f2dd0785ccf3c728f0e/eng/CodeStyle.targets).

Oh, thank you! 
We should move the versions to eng\Versions.props so we have all versions in the same place.

**Comment by elachlan** [comment]
> > The version that we use is in [`be7b036`/eng/CodeStyle.targets](https://github.com/dotnet/winforms/blob/be7b0368d986b6220d259f2dd0785ccf3c728f0e/eng/CodeStyle.targets).
> 
> Oh, thank you! (I should know these things, shouldn't I? :)) We should move the versions to eng\Versions.props so we have all versions in the same place.

Already done :)

**Comment by RussKie** [comment]
> I can sync with runtime if you want? Or did the team want to do that?

We take all the help we can from the community :)

**Comment by RussKie** [comment]
Do you mind updating ` <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.354" PrivateAssets="all" />` as well?

File: eng/CodeStyle.targets [hunk 1]
```diff
@@ -16,11 +16,11 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="6.0.0" PrivateAssets="all" />
+    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="$(MicrosoftCodeAnalysisNetAnalyzers)" PrivateAssets="all" />
```
Comment (RussKie) [inline eng/CodeStyle.targets:19]:
```suggestion
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="$(MicrosoftCodeAnalysisNetAnalyzersVersion)" PrivateAssets="all" />
```

File: eng/CodeStyle.targets [hunk 2]
```diff
@@ -16,11 +16,11 @@
...
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="DotNetAnalyzers.DocumentationAnalyzers" Version="1.0.0-beta.59" PrivateAssets="all" />
+    <PackageReference Include="DotNetAnalyzers.DocumentationAnalyzers" Version="$(DotNetAnalyzersDocumentationAnalyzers)" PrivateAssets="all" />
```
Comment (RussKie) [inline eng/CodeStyle.targets:23]:
```suggestion
    <PackageReference Include="DotNetAnalyzers.DocumentationAnalyzers" Version="$(DotNetAnalyzersDocumentationAnalyzersVersion)" PrivateAssets="all" />
```

File: eng/Versions.props [hunk 3]
```diff
@@ -91,6 +91,8 @@
     <MicrosoftCodeAnalysisVisualBasicWorkspacesVersion>4.0.0-2.final</MicrosoftCodeAnalysisVisualBasicWorkspacesVersion>
     <MicrosoftCodeAnalysisPackagesVersion>1.1.2-beta1.22415.2</MicrosoftCodeAnalysisPackagesVersion>
     <MicrosoftCodeAnalysisPublicApiAnalyzers>3.3.0-beta1.final</MicrosoftCodeAnalysisPublicApiAnalyzers>
```
Comment (RussKie) [inline eng/Versions.props:93]: Looks like we misnamed this too, it should have been
```suggestion
    <MicrosoftCodeAnalysisPublicApiAnalyzersVersion>3.3.0-beta1.final</MicrosoftCodeAnalysisPublicApiAnalyzersVersion>
```

File: eng/CodeAnalysis.src.globalconfig [hunk 4]
```diff
@@ -1,1430 +1,1744 @@
 is_global = true
 
-# Analyzer threw an exception
+# AD0001: Analyzer threw an exception
 dotnet_diagnostic.AD0001.severity = suggestion
```
Comment (RussKie) [inline eng/CodeAnalysis.src.globalconfig:4]: 👍

File: eng/CodeAnalysis.test.globalconfig [hunk 5]
```diff
@@ -1,1429 +1,1882 @@
 is_global = true
 
-# Analyzer threw an exception
-dotnet_diagnostic.AD0001.severity = suggestion
+# AD0001: Analyzer threw an exception
+dotnet_diagnostic.AD0001.severity = none
```
Comment (RussKie) [inline eng/CodeAnalysis.test.globalconfig:4]: Let's not reduce the existing levels. Applies to other instances.

Comment (elachlan) [reply eng/CodeAnalysis.test.globalconfig]: Done.

**Review by RussKie — APPROVED** [review]

**Comment by elachlan** [comment]
@RussKie only issue to resolve now is why the CA1852 doesn't detect the private classes we found in the other PR.

Does winforms use [InternalsVisibleToAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=net-6.0)?

**Comment by RussKie** [comment]
> Does winforms use [InternalsVisibleToAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=net-6.0)?

Yes, quite extensively: https://github.com/dotnet/winforms/search?q=InternalsVisibleTo

**Comment by elachlan** [comment]
Comment from the Analyzer code:
`// To avoid false positives, as with CA1812 (avoid uninstantiated internal classes), skip any assemblies with InternalsVisibleTo.`

The projects has the following and so is excluded:
https://github.com/dotnet/winforms/blob/6e32623149a53fba0311198eb577bce685c6c9d0/src/System.Windows.Forms.Design/src/Properties/AssemblyInfo.cs#L9-L13

We can continue to use it, but it won't find anything on projects using `InternalsVisibleTo`.

**Comment by elachlan** [comment]
An escape hatch for this behaviour is being tracked at https://github.com/dotnet/roslyn-analyzers/issues/5973

**Comment by RussKie** [comment]
Awesome! Thank you heaps for taking the lead on this.

**Pull request merged**